sklearn/cluster/optics_.py

```diff
@@ -682,18 +743,23 @@ def _cluster_tree(node, parent_node, local_maxima_points,
     node_list.append((node_1, local_max_1))
     node_list.append((node_2, local_max_2))
 
+    # if local maxima are sorted by their reachability, doesn't it
+    # mean we can stop? Why do we continue?
     if reachability_plot[s] < significant_min:
         node.assign_split_point(-1)
         # if split_point is not significant, ignore this split and continue
-        _cluster_tree(node, parent_node, local_maxima_points,
-                      reachability_plot, reachability_ordering,
-                      min_cluster_size, maxima_ratio, rejection_ratio,
-                      similarity_threshold, significant_min)
+        #_cluster_tree(node, parent_node, local_maxima_points,
+        #              reachability_plot, core_distances_plot,
+        #              ordering,
+        #              min_cluster_size, maxima_ratio, rejection_ratio,
+        #              similarity_threshold, significant_min)
         return
 
     # only check a certain ratio of points in the child
     # nodes formed to the left and right of the maxima
     # ...should check_ratio be a user settable parameter?
+    # NOTE: based on the description of maxima_ratio I'd guess this value
+    # should be 1, why is it less than 1?
```

Comment:
@espg, could you comment here, please?

Reply:
@espg, we're still not sure why the code doesn't include all the points here.

Reply:
Not sure, but I think it was to allow for outliers (i.e. if you have an inf or split point). The original code was hierarchical, so you could have nested clusters belonging to the same parent cluster. My guess (and it is a guess) is that the check ratio addresses that case for nested clusters.

Reply:
When I wrote this code, I recall it was because I was dealing with many closely nested candidate clusters, where the higher-level candidates in the cluster tree were being rejected for the lower-level candidates that were more tightly formed. I added this bit so I could tune the algorithm to be more inclusive of slightly further away points in the leaf clusters.

For the general purpose of having this in the library, this portion could probably be removed, as I don't think it's part of the original algorithm (@adrinjalali can confirm). This was the only place where I made any additional checks as compared to the paper.

Reply:
Yes, I read the paper yesterday ([here](https://link.springer.com/chapter/10.1007%2F3-540-36175-8_8) for reference), and this is not a part of the main algorithm. However, I wouldn't be opposed to including this as a parameter, with `1` as the default value as is in the paper. We may need to alter the algorithm a bit for the _split points_ issue anyway, and I know the other implementations (at least in R, although not of this algorithm, but the original OPTICS), don't strictly follow the publication and they may include some additions. But that's a different issue and I'm still working on how to articulate my thoughts on it the best I can.